plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/CloudLibraries.java

```diff
@@ -64,6 +87,63 @@ public static Library getLibrary(String id) {
     return libraries.get(id);
   }
 
+  private final static Supplier<List<Library>> clientApis = Suppliers.memoize(new Supplier<List<Library>>() {
+    public List<Library> get() {
+      Bundle bundle = FrameworkUtil.getBundle(CloudSdk.class);
+      URL url = bundle.getResource("/com/google/cloud/tools/libraries/libraries.json");
+      
+      try (InputStream in = url.openStream()) {
+        JsonReaderFactory factory = Json.createReaderFactory(null);
+        JsonReader reader = factory.createReader(in); 
+        JsonObject[] apis = reader.readArray().toArray(new JsonObject[0]); 
+        List<Library> clientApis = new ArrayList<>(apis.length);
+        for (JsonObject api : apis) {
+          String name = api.getString("name");
+          String id = api.getString("id");
+          Library library = new Library(id);
+          library.setGroup(CLIENT_APIS_GROUP);
+          library.setName(name);
+          JsonArray clients = api.getJsonArray("clients");
+          for (JsonObject client : clients.toArray(new JsonObject[0])) {
+            JsonString language = client.getJsonString("language");
+            if (language != null && "java".equals(language.getString())) {
+              String toolTip = client.getString("infotip");
+              library.setToolTip(toolTip);
+              JsonObject coordinates = client.getJsonObject("mavenCoordinates");
+              String groupId = coordinates.getString("groupId");
+              String artifactId = coordinates.getString("artifactId");
+              ArtifactVersion version =
+                  ArtifactRetriever.DEFAULT.getLatestArtifactVersion(groupId, artifactId);
```

Comment:
Getting the latest artifact version is potentially quite expensive, and we're going to do this on every library.  This doesn't feel like the right place.

Reply:
Where would you suggest doing it? It's only grabbed once and only requested when the first time the user opens the relevant panel. We could add a progress monitor.

Reply:
This could be a 30 page conference paper (which I've been threatening to submit for years now; maybe one of these days I'll actually write it); but tldr; GSON is, like most data binding products, brittle and unreliable. We don't want to serialize and unserialize objects. We want to query a data store for specific information we care about. 

Yes, this would work better if we had an actual query language such as XPath or if Java were more weakly typed, but GSON doesn't address either of those issues.

Reply:
We'd want it done on demand, no?  The user doesn't need the version number at selection time (if I'm looking for Cloud Datastore, then I won't be interested that Cloud Translate is version 1.22.3453).